src/common/key_value_database/platform/lmdb/lmdb.cpp

```diff
@@ -79,34 +89,65 @@ error::Error LmdbTransaction::Remove(const string &key) {
 	}
 }
 
-KeyValueDatabaseLmdb::KeyValueDatabaseLmdb() :
-	env_ {make_unique<lmdb::env>(lmdb::env::create())} {
+KeyValueDatabaseLmdb::KeyValueDatabaseLmdb() {
 }
 
 KeyValueDatabaseLmdb::~KeyValueDatabaseLmdb() {
 	Close();
 }
 
 error::Error KeyValueDatabaseLmdb::Open(const string &path) {
+	return OpenInternal(path, true);
+}
+
+error::Error KeyValueDatabaseLmdb::OpenInternal(const string &path, bool try_recovery) {
 	Close();
 
+	env_ = make_unique<lmdb::env>(lmdb::env::create());
+
 	try {
 		env_->open(path.c_str(), MDB_NOSUBDIR, 0600);
 	} catch (std::runtime_error &e) {
-		return MakeError(LmdbError, e.what());
+		auto err {MakeError(LmdbError, e.what()).WithContext("Opening LMDB database failed")};
+
+		if (not try_recovery) {
+			env_.reset();
+			return err;
+		}
+
+		try {
+			if (not fs::exists(path)) {
+				env_.reset();
+				return err;
+			}
+
+			log::Warning(
+				"Failure to open database. Attempting to recover by resetting. The original database can be found at `"
+				+ path + kBrokenSuffix + "`");
+
+			fs::rename(path, path + kBrokenSuffix);
+
+			auto err2 = OpenInternal(path, false);
+			if (err2 != error::NoError) {
+				return err.FollowedBy(err2);
+			}
+			return error::NoError;
+		} catch (fs::filesystem_error &e) {
+			env_.reset();
+			return err.FollowedBy(error::Error(e.code().default_error_condition(), e.what())
+									  .WithContext("Opening LMDB database failed"));
```

Comment:
Shouldn't this be also `LmdbError`?
```suggestion
			return err.FollowedBy(MakeError(LmdbError, e.what())
									  .WithContext("Opening LMDB database failed"));
```

Reply:
The error doesn't come from LMDB though, it comes from our file manipulation, outside of the database, so I thought it's better to be explicit about that.

Strictly speaking the exception category below (which I think cannot happen, but I didn't want to risk not having it), is not an `LmdbError` either, but I couldn't find a better one to give it, since `runtime_error` does not contain `code`. But maybe it's better to use `error::GenericError` there, since we have no idea where it comes from then. I'll do that, unless you object.

Reply:
Thanks. All things considered, it is fine as it is :+1: